cluster-autoscaler/cloudprovider/gce/cache.go

```diff
@@ -0,0 +1,258 @@
+	machinesCache   map[MachineTypeKey]*gce.MachineType
+	// Freshness timestamps.
+	machinesCacheLastRefresh time.Time
+	// Locks. Rules of locking:
+	// - migsMutex protects only migs.
+	// - cacheMutex protects migCache, resourceLimiter and machinesCache.
+	// - methods that take cacheMutex may call methods that take migsMutex.
+	// - methods that take migsMutex may not call methods that take cacheMutex.
+	cacheMutex sync.Mutex
+	migsMutex  sync.Mutex
+	// Service used to refresh cache.
+	GceService AutoscalingGceClient
+}
+
+// MigInformation is a wrapper for Mig.
+type MigInformation struct {
+	Config   Mig
+	Basename string
+}
+
+// MachineTypeKey is used to identify MachineType.
+type MachineTypeKey struct {
+	Zone        string
+	MachineType string
+}
+
+// NewGceCache creates empty GceCache.
+func NewGceCache(gceService AutoscalingGceClient) GceCache {
+	return GceCache{
+		migs:          []*MigInformation{},
+		migCache:      map[GceRef]Mig{},
+		machinesCache: map[MachineTypeKey]*gce.MachineType{},
+		GceService:    gceService,
+	}
+}
+
+//  Methods locking on migsMutex.
+
+// RegisterMig returns true if the node group wasn't in cache before, or its config was updated.
+func (gc *GceCache) RegisterMig(mig Mig) bool {
+	gc.migsMutex.Lock()
+	defer gc.migsMutex.Unlock()
+
+	for i := range gc.migs {
+		if oldMig := gc.migs[i].Config; oldMig.GceRef() == mig.GceRef() {
+			if !reflect.DeepEqual(oldMig, mig) {
+				gc.migs[i].Config = mig
+				glog.V(4).Infof("Updated Mig %s", mig.GceRef().String())
+				return true
+			}
+			return false
+		}
+	}
+
+	glog.V(1).Infof("Registering %s", mig.GceRef().String())
+	gc.migs = append(gc.migs, &MigInformation{
+		Config: mig,
+	})
+	return true
+}
+
+// UnregisterMig returns true if the node group has been removed, and false if it was alredy missing from cache.
+func (gc *GceCache) UnregisterMig(toBeRemoved Mig) bool {
+	gc.migsMutex.Lock()
+	defer gc.migsMutex.Unlock()
+
+	newMigs := make([]*MigInformation, 0, len(gc.migs))
+	found := false
+	for _, mig := range gc.migs {
+		if mig.Config.GceRef() == toBeRemoved.GceRef() {
+			glog.V(1).Infof("Unregistered Mig %s", toBeRemoved.GceRef().String())
+			found = true
+		} else {
+			newMigs = append(newMigs, mig)
+		}
+	}
+	gc.migs = newMigs
+	return found
+}
+
+// GetMigs returns a copy of migs list.
+func (gc *GceCache) GetMigs() []*MigInformation {
+	gc.migsMutex.Lock()
+	defer gc.migsMutex.Unlock()
+
+	migs := make([]*MigInformation, 0, len(gc.migs))
+	for _, mig := range gc.migs {
+		migs = append(migs, &MigInformation{
+			Basename: mig.Basename,
+			Config:   mig.Config,
+		})
+	}
+	return migs
+}
+
+func (gc *GceCache) updateMigBasename(ref GceRef, basename string) {
+	gc.migsMutex.Lock()
+	defer gc.migsMutex.Unlock()
+
+	for _, mig := range gc.migs {
+		if mig.Config.GceRef() == ref {
+			mig.Basename = basename
+		}
+	}
+}
+
+// Methods locking on cacheMutex.
+
+// GetMigForInstance returns MigConfig of the given Instance from cache.
```

Comment:
I realize this is a copy-paste, but this comment is really bad. MigConfig makes it look like a name, but we don't have anything named that. Also what is mig config of instance?

Reply:
Done